src/lib.rs

```diff
@@ -14,75 +21,88 @@ pub trait OwnedCharsExt {
 
 impl OwnedCharsExt for String {
     fn into_chars(self) -> OwnedChars {
-        OwnedChars { s: self, i: 0 }
+        OwnedChars::from_string(self)
     }
 
     fn into_char_indices(self) -> OwnedCharIndices {
-        OwnedCharIndices { s: self, i: 0 }
+        OwnedCharIndices::from_string(self)
     }
 }
 
 /// Iterator over the chars of a string (the string is owned by the iterator)
-#[derive(Clone, Debug)]
+#[derive(Debug)]
 pub struct OwnedChars {
     s: String,
-    i: usize,
+    i: Chars<'static>,
 }
 
 /// Iterator over the chars of a string and their indices (the string is owned by the iterator)
-#[derive(Clone, Debug)]
+#[derive(Debug)]
 pub struct OwnedCharIndices {
     s: String,
-    i: usize,
+    i: CharIndices<'static>,
 }
 
 macro_rules! impls {
-    ($s:ident) => {
-        impl $s {
+    ($struct:ident, $method: ident, $item: ty) => {
+        impl $struct {
+            /// Create Self from a String, moving the String into Self
+            fn from_string(s: String) -> Self {
+                // First, move the string
+                let mut owned = $struct {
+                    s: s,
+                    i: unsafe { uninitialized() }
+                };
+
+                // Then, we can call .chars, which with have the same lifetime
+                // of the owner. We need the transmute to "widen" the lifetime
+                // into 'static which would allow us to store it in the owner.
+                owned.i = unsafe { transmute(owned.s.$method()) };
```

Comment:
I'd prefer to write it as `transmute::<Chars, Chars<'static>(...)` just to be explicit that it's only changing the lifetime, but that'd require an extra parameter to the macro...